File: packages/share/src/__tests__/ShareModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { ShareModule } from '../ShareModule.js';

describe('ShareModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let share: ShareModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    share = new ShareModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  describe('share', () => {
    it('sends share request and returns result', async () => {
      const promise = share.share({ text: 'Hello', url: 'https://example.com', title: 'Test' });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('share');
      expect(sent.action).toBe('share');
      expect(sent.payload).toEqual({ text: 'Hello', url: 'https://example.com', title: 'Test' });

      respondSuccess({ success: true });
      const result = await promise;
      expect(result.success).toBe(true);
    });
  });

  describe('shareFile', () => {
    it('sends share file request', async () => {
      const promise = share.shareFile({ filePath: '/tmp/file.pdf', mimeType: 'application/pdf' });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.action).toBe('shareFile');
      respondSuccess({ success: true });
      const result = await promise;
      expect(result.success).toBe(true);
    });
  });

  describe('copyToClipboard', () => {
    it('sends copy to clipboard command', () => {
      share.copyToClipboard({ text: 'copied text' });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.action).toBe('copyToClipboard');
      expect(sent.payload).toEqual({ text: 'copied text' });
    });
  });

  describe('readClipboard', () => {
    it('reads clipboard text', async () => {
      const promise = share.readClipboard();
      respondSuccess({ text: 'clipboard content' });
      const result = await promise;
      expect(result.text).toBe('clipboard content');
    });
  });

  describe('canShare', () => {
    it('checks share capability', async () => {
      const promise = share.canShare();
      respondSuccess({ canShare: true });
      const result = await promise;
      expect(result.canShare).toBe(true);
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = share.share({ text: 'test' });
      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Share failed' },
      };
      transport.simulateIncoming(JSON.stringify(response));
      await expect(promise).rejects.toThrow('Share failed');
    });
  });
});

File: android/share/src/test/kotlin/io/rynbridge/share/ShareModuleTest.kt
package io.rynbridge.share

import io.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class ShareModuleTest {

    @Test
    fun `share returns success`() = runTest {
        val provider = MockShareProvider()
        val module = ShareModule(provider)
        val handler = module.actions["share"]!!

        val result = handler(mapOf(
            "text" to BridgeValue.string("Hello"),
            "url" to BridgeValue.string("https://example.com"),
            "title" to BridgeValue.string("Title")
        ))
        assertEquals(true, result["success"]?.boolValue)
        assertEquals("Hello", provider.lastShareText)
        assertEquals("https://example.com", provider.lastShareUrl)
        assertEquals("Title", provider.lastShareTitle)
    }

    @Test
    fun `shareFile returns success`() = runTest {
        val provider = MockShareProvider()
        val module = ShareModule(provider)
        val handler = module.actions["shareFile"]!!

        val result = handler(mapOf(
            "filePath" to BridgeValue.string("/path/to/file.pdf"),
            "mimeType" to BridgeValue.string("application/pdf")
        ))
        assertEquals(true, result["success"]?.boolValue)
        assertEquals("/path/to/file.pdf", provider.lastFilePath)
        assertEquals("application/pdf", provider.lastMimeType)
    }

    @Test
    fun `copyToClipboard stores text`() = runTest {
        val provider = MockShareProvider()
        val module = ShareModule(provider)
        val handler = module.actions["copyToClipboard"]!!

        val result = handler(mapOf("text" to BridgeValue.string("copied text")))
        assertTrue(result.isEmpty())
        assertEquals("copied text", provider.lastClipboardText)
    }

    @Test
    fun `readClipboard returns text`() = runTest {
        val provider = MockShareProvider()
        val module = ShareModule(provider)
        val handler = module.actions["readClipboard"]!!

        val result = handler(emptyMap())
        assertEquals("clipboard content", result["text"]?.stringValue)
    }

    @Test
    fun `canShare returns true`() = runTest {
        val provider = MockShareProvider()
        val module = ShareModule(provider)
        val handler = module.actions["canShare"]!!

        val result = handler(emptyMap())
        assertEquals(true, result["canShare"]?.boolValue)
    }

    @Test
    fun `module name and version`() {
        val provider = MockShareProvider()
        val module = ShareModule(provider)
        assertEquals("share", module.name)
        assertEquals("0.1.0", module.version)
    }
}

private class MockShareProvider : ShareProvider {
    var lastShareText: String? = null
    var lastShareUrl: String? = null
    var lastShareTitle: String? = null
    var lastFilePath: String? = null
    var lastMimeType: String? = null
    var lastClipboardText: String? = null

    override suspend fun share(text: String?, url: String?, title: String?): Boolean {
        lastShareText = text
        lastShareUrl = url
        lastShareTitle = title
        return true
    }

    override suspend fun shareFile(filePath: String, mimeType: String): Boolean {
        lastFilePath = filePath
        lastMimeType = mimeType
        return true
    }

    override suspend fun copyToClipboard(text: String) {
        lastClipboardText = text
    }

    override suspend fun readClipboard(): String? = "clipboard content"

    override suspend fun canShare(): Boolean = true
}
